src/rcdgen/string_storage.cpp

```diff
@@ -7,107 +7,292 @@
 
 /** @file string_storage.cpp Code of the translation storage. */
 
+#include <fstream>
+#include <set>
+#include <string>
+
 #include "../stdafx.h"
+#include "../string_func.h"
 #include "ast.h"
 #include "nodes.h"
 #include "string_storage.h"
 
 StringsStorage _strings_storage; ///< Storage of all (translated) strings by key.
 
-StringsStorage::StringsStorage()
-{
-}
-
 /**
- * Examine the still available strings, and decide whether it needs to be split on key.
- * @param strs Strings to examine.
- * @return Whether there are at least two keys in the collection, and it needs to be split.
+ * Retrieve the stored bundle of string by key name.
+ * @param key Key to retrieve.
+ * @return Pointer to the bundle, or \c nullptr if not found.
  */
-static bool NeedsSplit(std::shared_ptr<StringsNode> strs)
+const StringBundle *StringsStorage::GetBundle(const std::string &key)
 {
-	std::string key;
-	int num_keys = 0;
-	for (auto &str : strs->strings) {
-		if (num_keys == 0) {
-			key = str.key;
-			num_keys++;
-			continue;
-		}
-		if (str.key == key) continue;
-		return true;
-	}
-	return false;
+	auto iter = this->bundles.find(key);
+	if (iter == this->bundles.end()) return nullptr;
+	return &iter->second;
 }
 
 /**
- * Add a collection of strings to the storage.
- * @param strs String collection to add.
- * @param pos Position of the collection.
+ * Parse the provided YAML file and store its translations.
+ * @param filename File to read from.
  */
-void StringsStorage::AddStrings(std::shared_ptr<StringsNode> strs, const Position &pos)
+void StringsStorage::ReadFromYAML(const char *filename)
```

Comment:
A function of 250 lines seems excessive.

Can't you break it into smaller pieces? Perhaps use regular expression matching?